exercises/bank-account/src/test/java/BankAccountTest.java

```diff
@@ -0,0 +1,201 @@
+
+        bankAccount.withdraw(5);
+    }
+
+    @Ignore
+    @Test
+    public void cannotWithdrawMoreMoneyThanYouHave() throws BankAccountActionInvalidException {
+        bankAccount.open();
+        bankAccount.deposit(6);
+
+        expectedException.expect(BankAccountActionInvalidException.class);
+        expectedException.expectMessage("Cannot withdraw more money than is currently in the account");
+
+        bankAccount.withdraw(7);
+    }
+
+    @Ignore
+    @Test
+    public void cannotDepositNegativeAmount() throws BankAccountActionInvalidException {
+        bankAccount.open();
+
+        expectedException.expect(BankAccountActionInvalidException.class);
+        expectedException.expectMessage("Cannot deposit or withdraw negative amount");
+
+        bankAccount.deposit(-1);
+    }
+
+    @Ignore
+    @Test
+    public void cannotWithdrawNegativeAmount() throws BankAccountActionInvalidException {
+        bankAccount.open();
+        bankAccount.deposit(105);
+
+        expectedException.expect(BankAccountActionInvalidException.class);
+        expectedException.expectMessage("Cannot deposit or withdraw negative amount");
+
+        bankAccount.withdraw(-5);
+    }
+
+    @Ignore
+    @Test
+    public void cannotGetBalanceOfClosedAccount() throws BankAccountActionInvalidException {
+        bankAccount.open();
+        bankAccount.deposit(10);
+        bankAccount.close();
+
+        expectedException.expect(BankAccountActionInvalidException.class);
+        expectedException.expectMessage("Account closed");
+
+        bankAccount.getBalance();
+    }
+
+    @Ignore
+    @Test
+    public void cannotDepositMoneyIntoClosedAccount() throws BankAccountActionInvalidException {
+        bankAccount.open();
+        bankAccount.close();
+
+        expectedException.expect(BankAccountActionInvalidException.class);
+        expectedException.expectMessage("Account closed");
+
+        bankAccount.deposit(5);
+    }
+
+    @Ignore
+    @Test
+    public void cannotWithdrawMoneyFromClosedAccount() throws BankAccountActionInvalidException {
+        bankAccount.open();
+        bankAccount.deposit(20);
+        bankAccount.close();
+
+        expectedException.expect(BankAccountActionInvalidException.class);
+        expectedException.expectMessage("Account closed");
+
+        bankAccount.withdraw(5);
+    }
+
+    @Ignore
+    @Test
+    public void bankAccountIsClosedBeforeItIsOpened() throws BankAccountActionInvalidException {
+        expectedException.expect(BankAccountActionInvalidException.class);
+        expectedException.expectMessage("Account closed");
+
+        bankAccount.getBalance();
+    }
+
+    @Ignore
+    @Test
+    public void canAdjustBalanceConcurrently() throws BankAccountActionInvalidException, InterruptedException {
+        bankAccount.open();
+        bankAccount.deposit(1000);
+
+        for (int i = 0; i < 10; i++) {
+            adjustBalanceConcurrently();
```

Comment:
Ah, I see from the hint it is not guaranteed to fail. Can we provide any recommendation as to how many times a user should run this test to be "reasonably certain" their implementation is threadsafe? Alternatively, can we inject a little more thread control into solutions to be able to pull the appropriate levers ourselves and guarantee a concurrent modification attempt?